### DISCUSSION TITLE
How to combat goto brought by compiler

### DISCUSSION THREAD

**Discussion by imethod** [Vector35/binaryninja-api#4657] [Q&A] [answered]
During compiler optimization
An identical piece of code
```
int fun(int n)
{
  if (n== 3)
  {
    return 5
  }
  if (n == 2)
  {
    return 5
  }
  if (n == 1)
  {
    return 1
  }
}
```
This is what decompilation tends to look like after it's compiled
```
int fun(int n)
{
  if (n== 3)
  {
    goto LABEL_1;
  }
  if (n == 2)
  {
    goto LABEL_1;
  }
  if (n == 1)
  {
    return 1
  }
LABEL_1:
  return 5
}
```
In some larger functions especially algorithmic functions，They will iterate over each other resulting in extremely poor decompilation
![image](https://github.com/Vector35/binaryninja-api/assets/109936536/57a79a96-9673-44e8-93cc-88256cfdb7ad)
Does ninja have any good ways to combat this optimization

**Comment by imethod** [comment]
I may have found a solution
https://github.com/fkie-cad/dewolf

**Comment by plafosse** [comment, marked as answer]
There isn't a way to improve the control flow recover other than fixing up switch statements. Its unfortunately a "you get what you get" at this point. Feel free to report bugs though as we may be able to fix things.

**Comment by imethod** [comment]
> There isn't a way to improve the control flow recover other than fixing up switch statements. Its unfortunately a "you get what you get" at this point. Feel free to report bugs though as we may be able to fix things.

https://net.cs.uni-bonn.de/fileadmin/ag/martini/Staff/yakdan/dream_ndss2015.pdf
This may be a new way to restore control flow recovery
